src/components/tl/ucp/tl_ucp_context.c

```diff
@@ -228,6 +276,267 @@ ucc_status_t ucc_tl_ucp_populate_rcache(void *addr, size_t length,
+            goto fail_mem_map;
+        }
+        max_pack_size += my_pack_sizes[i];
+    }
+
+    total_pack_size =
+        (size_t *)ucc_malloc(sizeof(size_t) * size, "total_pack_size");
+    if (NULL == total_pack_size) {
+        tl_error(ctx->super.super.lib, "failed to allocated %zu bytes",
+                 sizeof(size_t) * size);
+        ucc_status = UCC_ERR_NO_MEMORY;
+        goto fail_mem_map;
+    }
+
+    ucc_status = ucc_tl_ucp_ctx_exchange_data(&max_pack_size, total_pack_size,
+                                              sizeof(size_t), ctx, oob);
+    if (UCC_OK != ucc_status) {
+        tl_error(ctx->super.super.lib,
+                 "failure in oob exchange of packed UCP key sizes");
+        ucc_status = UCC_ERR_NO_MESSAGE;
+        goto fail_pack;
+    }
+
+    for (i = 0; i < size; i++) {
+        if (total_pack_size[i] > max_pack_size) {
+            max_pack_size = total_pack_size[i];
+        }
+    }
+
+    ucc_status =
+        ucc_tl_ucp_ctx_remote_pack(map, my_pack, my_pack_sizes, max_pack_size,
+                                   &total, &packed_data, ctx);
+    if (UCC_OK != ucc_status) {
+        tl_error(ctx->super.super.lib, "failed to pack remote data");
+        goto fail_pack;
+    }
+
+    global_pack_data = ucc_malloc(total * size, "ucp_global_packed_data");
+    if (NULL == global_pack_data) {
+        tl_error(ctx->super.super.lib, "failed to allocated %zu bytes",
+                 sizeof(size_t) * size);
+        ucc_status = UCC_ERR_NO_MEMORY;
+        goto fail_pack;
+    }
+
+    ucc_status = ucc_tl_ucp_ctx_exchange_data(packed_data, global_pack_data,
+                                              total, ctx, oob);
+    if (UCC_OK != ucc_status) {
+        tl_error(ctx->super.super.lib,
+                 "failure in oob exchange of packed data");
+        goto fail_pack_exchange;
+    }
+
+    for (i = 0; i < size; i++) {
+        void *    base      = PTR_OFFSET(global_pack_data, i * total);
+        uint64_t *rvas      = base;
+        uint64_t *lens      = PTR_OFFSET(base, sizeof(uint64_t) * nsegs);
+        uint64_t *key_sizes = PTR_OFFSET(base, sizeof(uint64_t) * nsegs * 2);
+        void *    key       = PTR_OFFSET(base, sizeof(uint64_t) * nsegs * 3);
+        size_t    offset    = 0;
+
+        if (i != rank) {
+            remote_info[i] = (ucc_tl_ucp_remote_info_t *)ucc_calloc(
+                nsegs, sizeof(ucc_tl_ucp_remote_info_t), "ucp_remote_info");
+            if (NULL == remote_info[i]) {
+                tl_error(ctx->super.super.lib, "failed to allocated %zu bytes",
+                         sizeof(size_t) * size);
+                ucc_status = UCC_ERR_NO_MEMORY;
+
+                goto fail_unpack;
+            }
+        }
+
+        for (seg = 0; seg < nsegs; seg++) {
+            remote_info[i][seg].va_base = (void *)rvas[seg];
+            remote_info[i][seg].len     = lens[seg];
+            remote_info[i][seg].packed_key =
```

Comment:
what is the reason to save key and unpack it later on demand? can we instead unpack all keys here?

Reply:
if i recall correct, at this point in the code, i do not believe we have any endpoints yet. So, cannot unpack yet..